### PR TITLE
mrc-1908 Expose bundle interface to orderlyweb 

### PR DESCRIPTION
**Pull request by mwoodbri**
OrderlyWeb API support for [orderly bundles](https://www.vaccineimpact.org/orderly/articles/bundles.html). This enables the following:

1. Pack:
```
curl -H "Authorization: Bearer …" -H "Content-Type: application/json" -d '{"parameter": "value"}' -H "Accept: application/zip" -o bundle.zip localhost:8888/api/v1/bundle/pack/myreport/
```

2. Run:
```
R -e 'orderly::orderly_bundle_run("bundle.zip", workdir="/tmp")'
```

3. Import:
```
curl -H "Authorization: Bearer …" -H "Accept: application/json" —data-binary @/tmp/20201124-145140-a76d36c3.zip localhost:8888/api/v1/bundle/import/
```

Notes and outstanding issues:

- Both `pack` and `import` require `reports.run` privileges. The former is a read-only operation but is a precursor to the latter, which updates the repository in a way analogous to running a report. Rich has confirmed this approach is appropriate.
- All pre-existing interactions with orderly.server involve sending and receiving gzip-encoded JSON. The `/bundle/pack/:name` and `/bundle/import` endpoints respectively return and accept `application/zip` binary data respectively. This means that the `BundleController` is in some ways inconsistent with other controllers. In particular it uses OkHttp rather than khttp as the latter erroneously hard-codes the content type (see issues [48](https://github.com/ascclemens/khttp/issues/48), [58](https://github.com/ascclemens/khttp/issues/58)) for non-JSON non-multipart requests. For now, use of OkHttp is confined to these new orderly.server endpoints but given that khttp is effectively unmaintained I would suggest replacing it with OkHttp (which is already use for interacting with the Montagu API). I think this would merit its own ticket, but would allow some of this inconsistency to be resolved (and the same HTTP client abstractions to again be used across all endpoints).
- Bundles can only be imported once, have to be executed in R, and contain runtime data (e.g. checksums). This makes integration testing tricky because importing pre-completed bundles requires the deletion of any corresponding existing report version - which isn't supported by orderly.server (by design). My current approach is to snapshot and restore the database and delete the archive from disk. I can't immediately see a way round this, short of spinning up an R process (potentially in the orderly.server container?) to produce a novel bundle.
- A related issue is that deletion depends on knowing `orderly.root` but this is set differently in tests from what you may have for run-dependencies.sh. So tests will typically run in CI but not on your local machine. This needs to be resolved by  changing the way the the configuration works, or (less ideally) by only running the test in CI. Thoughts on this appreciated.
- A pre-completed bundle is required for testing. This is currently in `src/app/src/test/resources`, which has required a change to the way resources are copied into the build - the current step tries to expand any variables in the copied files, which isn't valid for binaries.
- The incoming and outgoing zip files are currently buffered in memory rather than being streamed. This is more straightforward to implement and easier to test but could be an issue if bundles are expected to be large.
- In `BundleController.pack()` there appears to be an assumption that JSON data specifying report parameters is a map containing `String` values. This assumption is common with the other endpoints related to orderly.server but having tested it there doesn't seem to be a problem passing numbers as parameters when packing reports, so I assume I've misunderstood something.
- IntelliJ is warning about that `context.params(":name")` can never be null. This is a result of `ActionContext.params()` having type `Map<String, String>`. I think this should be `params(): Map<String, String?>` but this would break at least one assumption elsewhere in the code.
- As noted in the ticket, orderly.server currently accepts and sets the `application/octet-stream` content type. This will change to the more correct `application/zip` when VIMC-4388 is closed. At this point the tests will fail and the `TODO` comments in `BundleController` can be resolved.

### CODE REVIEW

**Comment by hillalex** [comment]
> * All pre-existing interactions with orderly.server involve sending and receiving gzip-encoded JSON. The `/bundle/pack/:name` and `/bundle/import` endpoints respectively return and accept `application/zip` binary data respectively. This means that the `BundleController` is in some ways inconsistent with other controllers. In particular it uses OkHttp rather than khttp as the latter erroneously hard-codes the content type (see issues [48](https://github.com/ascclemens/khttp/issues/48), [58](https://github.com/ascclemens/khttp/issues/58)) for non-JSON non-multipart requests. For now, use of OkHttp is confined to these new orderly.server endpoints but given that khttp is effectively unmaintained I would suggest replacing it with OkHttp (which is already use for interacting with the Montagu API). I think this would merit its own ticket, but would allow some of this inconsistency to be resolved (and the same HTTP client abstractions to again be used across all endpoints).

I would be in favour of migrating `OrderlyServer` and the test's `RequestHelper` classes to OkHttp as a precursor to this branch, so that we don't end with inconsistencies in the codebase. But in particular I'd like all logic that interfaces with orderly.server to be encapsulated in the `OrderlyServer` class (so at the very least create methods for sending/receiving binary data on the `OrderlyServerAPI` interface/ in the `OrderlyServer` class that use OkHttp under the hood.)

> * Bundles can only be imported once, have to be executed in R, and contain runtime data (e.g. checksums). This makes integration testing tricky because importing pre-completed bundles requires the deletion of any corresponding existing report version - which isn't supported by orderly.server (by design). My current approach is to snapshot and restore the database and delete the archive from disk. I can't immediately see a way round this, short of spinning up an R process (potentially in the orderly.server container?) to produce a novel bundle.

We already [snapshot and restore the db as part of integration testing](https://github.com/vimc/orderly-web/blob/master/src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/IntegrationTest.kt#L44) but I think that in this case, the problem is that orderly.server is talking to the source db, rather than the test db. Still, I think this logic could be pulled into the `IntegrationTest @Before` and `@After` hooks - save out `db.template` into a third location before the tests run, then restore `db.template` from that location afterwards?

> * A related issue is that deletion depends on knowing `orderly.root` but this is set differently in tests from what you may have for run-dependencies.sh. So tests will typically run in CI but not on your local machine. This needs to be resolved by  changing the way the the configuration works, or (less ideally) by only running the test in CI. Thoughts on this appreciated.

The orderly demo data is a bit of a pain. We should really have one single source of demo data in the app. I think this ticket will be a step in the right direction: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-251 In the meantime, can we get the local run-dependencies script to use the same location as the tests?

> * The incoming and outgoing zip files are currently buffered in memory rather than being streamed. This is more straightforward to implement and easier to test but could be an issue if bundles are expected to be large.

:+1:  seems reasonable to go with the simpler solution for now and see how it performs. 

> * IntelliJ is warning about that `context.params(":name")` can never be null. This is a result of `ActionContext.params()` having type `Map<String, String>`. I think this should be `params(): Map<String, String?>` but this would break at least one assumption elsewhere in the code.

This seems like an oversight in the existing code - it would be nice if  `ActionContext.params()` returned `Map<String, String?>` (I don't think this method is actually ever referenced outside of the `DirectActionContext` class so this should be trivial to change?) and `ActionContext.params(key: String)` remains returning `String` but in the implementation we check for null and throw an error, so:

```
 override fun params(): Map<String, String?> = request.params()
 override fun params(key: String): String = request.params(key)?: throw MissingRequiredParameterError(key)
```

Seems to me like this change could be made without breaking any existing code.

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                    context,
+                    AppConfig(),
+                    OkHttpClient()
+            )
+
+    fun pack(): Boolean
+    {
+        val name = context.params(":name") ?: throw MissingParameterError("name")
+
+        val url = appConfig["orderly.server"] + "/v1/bundle/pack/$name" + (if (context.queryString() != null) "?" + context.queryString() else "")
```
Comment (hillalex) [inline src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt:33]: As mentioned, can we move all this logic to the `OrderlyServer` class?

A pedantic style point, our convention is to wrap all conditionals in braces, but I would make this more kotlin-esque as follows:

`val url = "${appConfig["orderly.server"]}/v1/bundle/pack/$name?${context.queryString()?: "" }"`

(^ does add an extra "?" even when no query string, but that seems harmless enough)

**Review by EmmaLRussell — COMMENTED** [review]
This is probably a daft question, but is there a reason why we're not providing the orderly_bundle_run endpoint too?

>given that khttp is effectively unmaintained I would suggest replacing it with OkHttp (which is already use for interacting with the Montagu API

Sounds good to me, yes definitely one for another ticket. 

I think the issue with `context.params(":name")` is because that's using the `params(key: String)` method. `context.params()[":name"]` should allow null to be potentially returned if the key doesn't exist I think.

**Comment by EmmaLRussell** [comment]
Oh sorry, clashing reviews

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            this(
+                    context,
+                    AppConfig(),
+                    OkHttpClient()
+            )
+
+    fun pack(): Boolean {
+        val name = context.params(":name") ?: throw MissingParameterError("name")
+
+        val url = appConfig["orderly.server"] + "/v1/bundle/pack/$name" + (if (context.queryString() != null) "?" + context.queryString() else "")
```
[the remark this answers is not in the document]
Comment (hillalex) [reply src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt:33]: I've just added some docs on code styles, Emma please edit if there's anything I've forgotten :) https://mrc-ide.myjetbrains.com/youtrack/articles/RESIDE-A-8/Code-style-conventions

[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt]: Done in 23c5b07

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        val body = getResource("$reportVersion.zip").readBytes()
+        val request = Request.Builder()
+                .url("${apiRequestHelper.baseUrl}/bundle/import/")
+                .header("Authorization", "Bearer $token")
+                .post(body.toRequestBody("application/zip".toMediaType()))
+                .build()
+        val response = OkHttpClient().newCall(request).execute()
+        assertThat(response.code).isEqualTo(200)
+
+        // Restore orderly repository state
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt]: Done - though doesn't resolve question over backup/restore of orderly.server's store, which may differ from OrderlyWeb's). See [Testing](https://mrc-ide.myjetbrains.com/youtrack/articles/mrc-A-15/Testing) for possible solutions.

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        val response = apiRequestHelper.post("/bundle/pack/minimal/", emptyMap(), ContentTypes.zip, fakeGlobalReportReviewer())
+        assertThat(response.statusCode).isEqualTo(200)
+        assertThat(response.headers["content-type"]).isEqualTo("application/zip")
+        val contents = getZipEntries(response).map { it.split('/', limit = 2).last() }
+        assertThat(contents).contains("meta/manifest.rds", "pack/orderly.yml")
+    }
+
+    @Test
+    fun `packs report fails without required permission`() {
+        val response = apiRequestHelper.post("/bundle/pack/minimal/", emptyMap(), ContentTypes.zip)
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt]: Done - thanks!

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        val reportVersion = "20201126-153124-9a32811a"
+        val orderlyRoot = AppConfig()["orderly.root"]
+
+        // Backup orderly repository state
+        val dbPath = Paths.get("${orderlyRoot}orderly.sqlite")
+        val dbContent = Files.readAllBytes(dbPath)
+
+        val token = WebTokenHelper.instance.issuer.generateBearerToken(fakeGlobalReportReviewer())
+        val body = getResource("$reportVersion.zip").readBytes()
+        val request = Request.Builder()
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt:87]: Prerequisite for this is to remove final usages of khttp (from helpers) - but I think this merits another maintenance ticket. See [mrc-2024](https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2024).

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    @Test
+    fun `packs report fails without required permission`() {
+        val response = apiRequestHelper.post("/bundle/pack/minimal/", emptyMap(), ContentTypes.zip)
+        assertThat(response.statusCode).isEqualTo(403)
+    }
+
+    @Test
+    fun `imports report()`() {
+        val reportName = "minimal"
+        val reportVersion = "20201126-153124-9a32811a"
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt]: Have added a comment - this ID isn't dependent on contents of demo database - it just needs to be unique (though this is dependent on current orderly.server implementation)

**Comment by mwoodbri** [comment]
@EmmaLRussell Neither Travis or Codecov seem to be triggered for most recent commits - have also tried to do manually with no luck. Any ideas welcome!

**Comment by EmmaLRussell** [comment]
Maybe because of Travis's inexplicable new pricing plan? I'm seeing *"Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting."* on the travis page for the repo, though it has been running my webapps builds ok today. Will see if I can kick it into action. 

> @EmmaLRussell Neither Travis or Codecov seem to be triggered for most recent commits - have also tried to do manually with no luck. Any ideas welcome!

**Comment by EmmaLRussell** [comment]
Nope, didn't work. The tooltip on the rerun button in the build history also says that builds have been disabled on the repo due to a negative credit balance..  So we should endeavour to get codecov into buildkite asap..

> Maybe because of Travis's inexplicable new pricing plan? I'm seeing _"Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting."_ on the travis page for the repo, though it has been running my webapps builds ok today. Will see if I can kick it into action.

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/OrderlyServerTests.kt [hunk 7]
```diff
@@ -97,6 +135,17 @@ class OrderlyServerTests
         )
     }
 
+    @Test
+    fun `throws error on failure`()
```
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/OrderlyServerTests.kt:139]: I think this is a partial duplicate of `throws error on failure` which I guess we don't need any more. But are we missing a test that it *doesn't* throw if `.throwOnError()` not called?

**Review by EmmaLRussell — COMMENTED** [review]
All looks good to me, just a tiny query about one of the tests.

**Review by EmmaLRussell — APPROVED** [review]
Thanks for that!

**Pull request merged**